source/event_stream_rpc_server.c

```diff
@@ -0,0 +1,914 @@
+                                          "stream-id arrived that was lower than the last seen stream-id.\"; }");
+
+static const struct aws_byte_cursor s_invalid_new_client_stream_id_error =
+    AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("{ \"message\": \"stream-id values must be monotonically incrementing. A new "
+                                          "stream-id arrived that was incremented by more than 1.\"; }");
+
+static const struct aws_byte_cursor s_missing_operation_name_error = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL(
+    "{ \"message\": \"The first message for on a non-zero :stream-id must contain an operation header value.\"; }");
+
+static const struct aws_byte_cursor s_invalid_message_type_error = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL(
+    "{ \"message\": \"an invalid value for message-type field was received.\"; }");
+
+static const struct aws_byte_cursor s_server_error = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL(
+    "{ \"message\": \"an error occurred on the server. This is not likely caused by your client.\"; }");
+
+static const struct aws_byte_cursor s_invalid_message_error = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL(
+    "{ \"message\": \"A message was received with missing required fields. Check that your client is sending at least, "
+    ":message-type, :message-flags, and :stream-id\"; }");
+
+static const struct aws_byte_cursor s_connect_not_completed_error = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL(
+    "{ \"message\": \"A CONNECT message must be received, and the CONNECT_ACK must be sent in response, before any "
+    "other message-types can be sent on this connection. In addition, only one CONNECT message is allowed on a "
+    "connection.\" }");
+
+static const uint32_t s_bit_scrambling_magic = 0x45d9f3bU;
+static const uint32_t s_bit_shift_magic = 16U;
+
+/* this is a repurposed hash function based on the technique in splitmix64. The magic number was a result of numerical
+ * analysis on maximum bit entropy. */
+static uint64_t s_hash_uint32(const void *to_hash) {
+    uint32_t int_to_hash = *(const uint32_t *)to_hash;
+    uint32_t hash = ((int_to_hash >> s_bit_shift_magic) ^ int_to_hash) * s_bit_scrambling_magic;
+    hash = ((hash >> s_bit_shift_magic) ^ hash) * s_bit_scrambling_magic;
+    hash = (hash >> s_bit_shift_magic) ^ hash;
+    return (uint64_t)hash;
+}
+
+static bool s_uint32_eq(const void *a, const void *b) {
+    return *(const uint32_t *)a == *(const uint32_t *)b;
+}
+
+struct aws_event_stream_rpc_server_listener {
+    struct aws_allocator *allocator;
+    struct aws_socket *listener;
+    struct aws_server_bootstrap *bootstrap;
+    struct aws_atomic_var ref_count;
+    aws_event_stream_rpc_server_on_new_connection_fn *on_new_connection;
+    aws_event_stream_rpc_server_on_connection_shutdown_fn *on_connection_shutdown;
+    aws_event_stream_rpc_server_on_listener_destroy_fn *on_destroy_callback;
+    size_t initial_window_size;
```

Comment:
Just offering flow-control over the socket read might not be sufficient, especially since this is a multi-stream protocol, to prevent the server from being overwhelmed. Consider:
- Control over the max concurrent streams?
  - Lets the server throttle the maximum amount of work it can handle at once. With the current design, I guess the server could just queue up requests and only actively "handle" a fixed number at a time. So I guess adding a limit to the protocol would just force the clients to queue requests, instead of forcing the server to queue requests... So not critical, just bringing up the design choices.
- Control over stream window-size, instead of (or in addition to) connection window-size?
  - Prevents head-of-line blocking. With only connection level flow-control, an operation that consumes bytes slowly will hog the available window, preventing other streams from receiving data. If we only had stream level flow-control, then we'd avoid head of line blocking because fast streams could continue flowing if a slow stream stalls. If we had both stream and connection level flow-control, a connection could also limit the total outstanding bytes. Not sure how valuable it is to have both, but stream control seems more important than connection control.
- Should "control" frames not count? Should only the payload count?
  - In HTTP/2, flow-control only applies to the payload of DATA frames, so control frames can always get through. With the current design, if an endpoint's window is full, it can't even receive PING frames.

Reply:
1. throttling would be handled by the user now, or later at a higher level. But the idea would be to either send an APPLICATION_ERROR on the new stream (after the first message arrives) with the throttled response, or a connection wide PROTOCOL_ERROR if the contract has been violated.

2. in current mode, manual window management is un-implemented, and nothing prevents us from augmenting the protocol in the future to include stream windowing. At the moment, I'm not particularly fond of the complexity this would add.

3. control frames should not count (though not implemented at the moment). This brings up an interesting point however about not relying on socket level windowing... my vote at the moment is to revisit this in the next iteration (when we enable manual window management).

Reply:
okay the backpressure setting is now removed